core/src/main/java/io/grpc/internal/ManagedChannelServiceConfig.java

```diff
@@ -162,28 +164,37 @@ static ManagedChannelServiceConfig fromServiceConfig(
             healthCheckingConfig);
   }
 
-  /**
-   * Returns the per-service configuration for the channel.
-   */
-  Map<String, MethodInfo> getServiceMap() {
-    return serviceMap;
-  }
-
   @Nullable
   Map<String, ?> getHealthCheckingConfig() {
     return healthCheckingConfig;
   }
 
   /**
-   * Returns the per-method configuration for the channel.
+   * Used as a fallback per-RPC config supplier when the attributes value of {@link
+   * InternalConfigSelector#KEY} is not available.
    */
-  Map<String, MethodInfo> getServiceMethodMap() {
-    return serviceMethodMap;
-  }
-
   @Nullable
-  MethodInfo getDefaultMethodConfig() {
-    return defaultMethodConfig;
+  InternalConfigSelector getDefaultConfigSelector() {
+    if (serviceMap.isEmpty() && serviceMethodMap.isEmpty() && defaultMethodConfig == null) {
+      return null;
+    }
+    return new InternalConfigSelector() {
+      @Override
+      public Result selectConfig(PickSubchannelArgs args) {
+        MethodInfo methodInfo = getMethodConfig(args.getMethodDescriptor());
+        return Result.newBuilder()
+            .setConfig(
+                new ManagedChannelServiceConfig(
```

Comment:
Why not use `this`? Was this from an earlier revision and wasn't updated?

Reply:
The new instance strips off all configs (e.g. lb config) other than method configs from `this`. The design seems to say the `Result.config` is per RPC config.

Reply:
It seems to say, "this is necessary." If we want to say, "only the per-RPC config will be used" then we can use a comment to do that.

I'm really not wild with calling getMethodConfig() here. We should just return `this` and have this be a trivial implementation.

Reply:
Now I put entire service config in `setConfig()`. Now that the config selector contains all the information of the ManagedConfigServiceConfig, I rename `ManagedConfigServiceConfig.getDefaultConfigSelector()` as `ManagedConfigServiceConfig.asConfigSelector()`.

Reply:
You removed the "return null when a config selector is unnecessary" optimization. That means now `configSelector.set(null)` is only used for shutdown when there are no calls and error cases. That's asking for bugs. Restore the return-null optimization or prevent the config selector from ever being null.

Reply:
Good point. Didn't think about it. Restored return-null optimization.